### PR TITLE
Sorting Params

### PR DESCRIPTION
**Pull request by JinKoro**

### CODE REVIEW

File: src/main/kotlin/dev/tmsoft/lib/params/filter/Router.kt [hunk 1]
```diff
@@ -32,14 +35,11 @@ private fun Filter.openApiType(): Type.Object {
...
+        if (key.contains("filter")) {
             val result = Regex("\\[(\\w+)\\]").findAll(key)
             val field: String = result.last().groupValues.last()
             parameters[field] = QueryConverter.convert(value)
         }
     }
     return PathValues(parameters)
 }
-
-private val parameterName: String
```
Comment (AndreiBaly) [inline src/main/kotlin/dev/tmsoft/lib/params/filter/Router.kt:44]: It's better to leave it as it was, just renamed to filterParameterName

File: src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+abstract class Sorting(val table: Table) {
+    private val fields: MutableList<Field> = mutableListOf()
+
+    fun fields(): List<Field> = fields
+
+    fun add(
+        name: String,
+        column: Column<*>? = null,
+        possibleValues: List<String>? = null
```
Comment (AndreiBaly) [inline src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt:20]: What is this parameter for?

File: src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    ): Field {
+        val tableColumn = column ?: table.columns.first { it.name == name }
+
+        val field = Field(
+            name,
+            { values ->
+                val sortOrder = (values.single() as SingleValue).value
+                    .replace(" ","")
```
Comment (AndreiBaly) [inline src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt:28]: Maybe this `replace` is not needed?

File: src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt (hunk 2 above)
Comment (JinKoro) [reply src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt:20]: for openApi -> swagger

File: src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt (hunk 3 above)
Comment (JinKoro) [reply src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt:28]: I think it give more path options. Because SortOrder.valuesOf("desc ") with whitespace doesn't work

Comment (shustrik) [reply src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt:28]: I think we should have method trim()

File: src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt (hunk 2 above)
Comment (shustrik) [reply src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt:20]: and what values are possible in this field? ASC, DESC?

Comment (JinKoro) [reply src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt:20]: ASC,
DESC,
ASC_NULLS_FIRST,
DESC_NULLS_FIRST,
ASC_NULLS_LAST,
DESC_NULLS_LAST
@shustrik

Comment (shustrik) [reply src/main/kotlin/dev/tmsoft/lib/params/sorting/Sorting.kt:20]: Ok, but it's a constant, I can't provide MY_SORT_ORDER

File: src/main/kotlin/dev/tmsoft/lib/openapi/OpenAPI.kt [hunk 4]
```diff
@@ -28,32 +28,44 @@ class OpenAPI(var host: String) {
             pathItemObject = PathItemObject()
             root.paths[path] = pathItemObject
         }
+        val updatedPathParams = responses.values
```
Comment (shustrik) [inline src/main/kotlin/dev/tmsoft/lib/openapi/OpenAPI.kt:31]: Remove this, open API shouldn't know about pages or filters or sorting

File: src/main/kotlin/dev/tmsoft/lib/ktor/Response.kt [hunk 5]
```diff
@@ -2,8 +2,8 @@
 
 package dev.tmsoft.lib.ktor
 
-import dev.tmsoft.lib.query.ContinuousList
-import dev.tmsoft.lib.query.ContinuousListSerializer
+import dev.tmsoft.lib.params.paging.ContinuousList
+import dev.tmsoft.lib.params.paging.ContinuousListSerializer
```
Comment (shustrik) [inline src/main/kotlin/dev/tmsoft/lib/ktor/Response.kt]: change package to listing, move query objects to exposed

File: src/main/kotlin/dev/tmsoft/lib/params/paging/ContinuousList.kt [hunk 6]
```diff
@@ -16,17 +18,29 @@ import org.jetbrains.exposed.sql.Query
 import org.jetbrains.exposed.sql.ResultRow
 import org.jetbrains.exposed.sql.ops.SingleValueInListOp
 
-fun <T> Query.toContinuousList(page: PagingParameters, effector: ResultRow.() -> T): ContinuousList<T> {
-    return toContinuousListBuilder(page) { this.map { effector(it) } }
+fun <T> Query.toContinuousList(
+    page: PagingParameters,
+    effector: ResultRow.() -> T,
+    sorting: Sorting? = null,
```
Comment (shustrik) [inline src/main/kotlin/dev/tmsoft/lib/params/paging/ContinuousList.kt]: sorting and sortingValues combine at the single object Sorting

File: src/main/kotlin/dev/tmsoft/lib/params/paging/sorting/Sorting.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+import dev.tmsoft.lib.params.Field
+import dev.tmsoft.lib.params.PathValues
+import dev.tmsoft.lib.params.SingleValue
+import org.jetbrains.exposed.sql.Column
+import org.jetbrains.exposed.sql.Query
+import org.jetbrains.exposed.sql.SortOrder
+import org.jetbrains.exposed.sql.Table
+
+abstract class Sorting(val table: Table) {
```
Comment (shustrik) [inline src/main/kotlin/dev/tmsoft/lib/params/paging/sorting/Sorting.kt]: remove

**Review by shustrik — CHANGES_REQUESTED** [review]
@AndreiBaly explain to you all comments

File: detekt-baseline.xml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" ?>
+<SmellBaseline>
+  <ManuallySuppressedIssues></ManuallySuppressedIssues>
+  <CurrentIssues>
+    <ID>SpreadOperator:ContinuousList.kt$(*orders)</ID>
```
Comment (shustrik) [inline detekt-baseline.xml:5]: remove baseline you should resolve all detekt issues

File: src/main/kotlin/dev/tmsoft/lib/query/paging/ContinuousList.kt [hunk 9]
```diff
@@ -16,28 +16,52 @@ import org.jetbrains.exposed.sql.Query
...
 }
 
 @JvmName("toContinuousListIterable")
-fun <T> Query.toContinuousList(page: PagingParameters, effector: Iterable<ResultRow>.() -> List<T>): ContinuousList<T> {
-    return toContinuousListBuilder(page) { effector() }
+fun <T> Query.toContinuousList(
+    page: PagingParameters,
+    effector: Iterable<ResultRow>.() -> List<T>,
+    sortingParameters: List<SortingParameter>? = null,
+    columns: List<Column<*>>? = null
```
Comment (shustrik) [inline src/main/kotlin/dev/tmsoft/lib/query/paging/ContinuousList.kt:33]: you can get columns from Query or  "this" context

File: src/main/kotlin/dev/tmsoft/lib/query/filter/Value.kt [hunk 10]
```diff
@@ -34,8 +47,6 @@ class SingleValue(val value: String) : Value() {
     @Suppress("UNCHECKED_CAST")
     private fun ExpressionWithColumnType<*>.expression(value: String): Op<Boolean> {
         return when (columnType) {
-            is EnumerationColumnType<*> -> EqOp(this, typedWrap(value))
```
Comment (shustrik) [inline src/main/kotlin/dev/tmsoft/lib/query/filter/Value.kt:37]: why did you remove these operations

**Review by shustrik — CHANGES_REQUESTED** [review]

**Review by shustrik — APPROVED** [review]
